source/vibe/core/sync.d

```diff
@@ -819,3 +819,469 @@ private struct TaskConditionImpl(bool INTERRUPTIBLE, LOCKABLE) {
+        }
+        else
+            m_counterMutex.lock();
+        
+        scope(exit)
+            m_counterMutex.unlock();
+        
+        //Log a debug statement for the attempt
+        version(RWMutexPrint)
+            printInfo!("checkCtrs",INTENT)();
+        
+        //Check if there's already an active writer
+        if (m_activeWriteLocks > 0)
+            return false;
+        
+        //If writers are preferred over readers, check whether there
+        //currently is a writer in the waiting queue and abort if
+        //that's the case.
+        static if (INTENT == LockingIntent.READ_ONLY)
+            if (m_policy.PREFER_WRITERS && m_waitingForWriteLock > 0)
+                return false;
+        
+        //If we are locking the mutex for writing, make sure that
+        //there's no reader active.
+        static if (INTENT == LockingIntent.READ_WRITE)
+            if (m_activeReadLocks > 0)
+                return false;
+        
+        //We can successfully acquire the lock!
+        //Log a debug statement for the success.
+        version(RWMutexPrint)
+            printInfo!("lock",INTENT)();
+        
+        //Increase the according counter 
+        //(number of active readers/writers)
+        //and return a success code.
+        activeCounter!INTENT += 1;
+        return true;
+    }
+    
+    /** Attempt to acquire the lock for a given intent.
+     *
+     *  Returns:
+     *      `true`, if the lock was successfully acquired;
+     *      `false` otherwise.
+     */
+    @trusted bool tryLock(LockingIntent INTENT)()
+    {
+        //Try to lock this mutex without waiting for the underlying
+        //TaskMutex - fail if it is already blocked.
+        return tryLock!(INTENT,false)();
+    }
+    
+    /** Acquire the lock for the given intent; yield and suspend until the lock has been acquired. */
+    @trusted void lock(LockingIntent INTENT)()
+    {
+        //Prepare a waiting action before the first
+        //`tryLock()` call in order to avoid a race
+        //condition that could lead to the queue notification
+        //not being fired.
+        auto count = emitCount!INTENT;
+        atomicOp!"+="(queueCounter!INTENT,1);
+        scope(exit)
+            atomicOp!"-="(queueCounter!INTENT,1);
+        
+        //Try to lock the mutex
+        auto locked = tryLock!(INTENT,true)();
+        if (locked)
+            return;
+        
+        //Retry until we successfully acquired the lock
+        while(!locked)
+        {
+            version(RWMutexPrint)
+                printInfo!("wait",INTENT)();
+            
+            count  = wait!INTENT(count);
+            locked = tryLock!(INTENT,true)();
+        }
+    }
+    
+    /** Unlock the mutex after a successful acquisition. */
+    @trusted void unlock(LockingIntent INTENT)()
+    {
+        version(RWMutexPrint)
+            printInfo!("unlock",INTENT)();
+        
+        debug assert(activeCounter!INTENT > 0);
```

Comment:
Not completely true, `assert` will be there when neither `-debug`, nor `-release` is used, but `debug` statements are only included if `-debug` is given. I usually use `debug assert` in performance critical places, so that it's possible to build a fast "release" build without resorting to the unsafe `-release` switch.